### PR TITLE
Added dsh-bio filter-bed as dsh_filterbed

### PR DESCRIPTION
**Pull request by heuermh**
WIP

### CODE REVIEW

File: software/dsh/splitbed/main.nf [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    conda (params.enable_conda ? "bioconda::dsh-bio=2.0" : null)
+
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/dsh-bio:2.0--0"
+    } else {
+        container "quay.io/biocontainers/dsh-bio:2.0--0"
+    }
+
+    input:
+    tuple val(meta), path(features)
```
Comment (drpatelh) [inline software/dsh/splitbed/main.nf:23]:
```suggestion
    tuple val(meta), path(bed)
```

File: software/dsh/splitbed/meta.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        type: file
+        description: Features in gzipped BED format to split
+        pattern: "*.{bed.gz}"
+output:
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - features:
```
Comment (drpatelh) [inline software/dsh/splitbed/meta.yml:50]:
```suggestion
    - bed:
```

File: software/dsh/filterbed/main.nf [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        container "https://depot.galaxyproject.org/singularity/dsh-bio:2.0--0"
+    } else {
+        container "quay.io/biocontainers/dsh-bio:2.0--0"
+    }
+
+    input:
+    tuple val(meta), path(features)
+
+    output:
+    tuple val(meta), path("*.bed.gz"), emit: features
```
Comment (drpatelh) [inline software/dsh/filterbed/main.nf:26]:
```suggestion
    tuple val(meta), path("*.bed.gz"), emit: bed
```

File: software/dsh/filterbed/main.nf [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    conda (params.enable_conda ? "bioconda::dsh-bio=2.0" : null)
+
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/dsh-bio:2.0--0"
+    } else {
+        container "quay.io/biocontainers/dsh-bio:2.0--0"
+    }
+
+    input:
+    tuple val(meta), path(features)
```
Comment (drpatelh) [inline software/dsh/filterbed/main.nf:23]:
```suggestion
    tuple val(meta), path(bed)
```

File: software/dsh/filterbed/main.nf [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    path "*.version.txt"             , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    """
+    dsh-filter-bed \\
+        $options.args \\
+        -i $features \\
```
Comment (drpatelh) [inline software/dsh/filterbed/main.nf:36]:
```suggestion
        -i $bed \\
```

File: software/dsh/filterbed/meta.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        description: |
+            Instead of directly downloading Singularity images for use with Singularity,
+            force the workflow to pull and convert Docker containers instead.
+input:
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - features:
```
Comment (drpatelh) [inline software/dsh/filterbed/meta.yml:40]:
```suggestion
    - bed:
```

File: software/dsh/filterbed/meta.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        type: file
+        description: Features in gzipped BED format
+        pattern: "*.{bed.gz}"
+output:
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - features:
```
Comment (drpatelh) [inline software/dsh/filterbed/meta.yml:50]:
```suggestion
    - bed:
```

File: software/dsh/splitbed/main.nf [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    """
+    dsh-split-bed \\
+        $options.args \\
+        -p $prefix \\
+        -s '.bed.gz' \\
+        -i $features
```
Comment (drpatelh) [inline software/dsh/splitbed/main.nf:38]:
```suggestion
        -i $bed
```

**Review by drpatelh — CHANGES_REQUESTED** [review]
Looks great @heuermh ! Thank you 😎 

Mainly just changing `features` -> `bed` in places to try and have more generic named input / outputs wherever possible.

**Comment by heuermh** [comment]
Thanks for the review, @drpatelh!

As discussed on Slack, I've given up here on supporting multiple compression types and default to gzip.  I am still thinking on whether it would be worth the mess of additional inputs & outputs.

Is it worth calling programs to get the version (e.g. https://github.com/nf-core/modules/pull/118/files#diff-78760d8d7b3805cfb74e63ef869f845c02dc6dea1a92e90b99ccca587ef78d04R39) or since we already know the version from the Bioconda URLs, better to output it directly?  Same goes for `getSoftwareName`.

File: software/dsh/filterbed/main.nf [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    tuple val(meta), path(bed)
+
+    output:
+    tuple val(meta), path("*.bed.gz"), emit: bed
+    path "*.version.txt"             , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
```
Comment (drpatelh) [inline software/dsh/filterbed/main.nf:32]:
```suggestion
```

File: software/dsh/filterbed/main.nf [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+process DSH_FILTERBED {
+    tag "${meta.id}"
+    label 'process_medium'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), publish_id:meta.id) }
+
+    conda (params.enable_conda ? "bioconda::dsh-bio=2.0" : null)
+
```
Comment (drpatelh) [inline software/dsh/filterbed/main.nf:15]:
```suggestion
```

**Review by drpatelh — APPROVED** [review]
Couple of minor things left and then feel free to merge @heuermh

**Comment by drpatelh** [comment]
You could check the file extension within the module and take the appropriate steps. I think its fine to enforce users to use zipped format for now though 👍🏽 The module can always be updated again if it becomes a problem.

The software version has to be obtained via a command because we can also have containers called [this](https://github.com/nf-core/modules/blob/29ce4e670f3da41273b0c6b78003b94a1c163626/software/bwa/mem/main.nf#L16)! Not all containers will be on Conda either...so things can start becoming quite inconsistent.

**Comment by drpatelh** [comment]
Thanks @heuermh ! Let's get this in before the weekend 🏖️

**Comment by heuermh** [comment]
Thanks, @drpatelh!  I have maybe a dozen more `dsh-*` modules to add, just wanted to establish the correct pattern before doing the others.

> The software version has to be obtained via a command because we can also have containers called this! Not all containers will be on Conda either...so things can start becoming quite inconsistent.

I see a performance issue with this for Spark-based jobs, even simply calling `--version` and exiting can distributed as a job out to the cluster, so round trip for this call could be 30 seconds or more.

**Pull request merged**